store/copr/batch_coprocessor.go

```diff
@@ -798,15 +781,99 @@ func failpointCheckWhichPolicy(act tiflashcompute.DispatchPolicy) {
 	})
 }
 
+func filterAllStoresAccordingToTiFlashReplicaRead(allStores []uint64, aliveStoreIDs map[uint64]struct{}, aliveStoreIDsInTiDBZone map[uint64]struct{}, isTiDBZoneSet bool, policy tiflash.ReplicaRead) (storesMatchedPolicy []uint64, needsCrossZoneAccess bool) {
+	if policy == tiflash.AllReplicas || !isTiDBZoneSet {
+		return allStores, false
+	}
+	// Check whether exists available stores in TiDB zone. If so, we only need to access TiFlash stores in TiDB zone.
+	for _, id := range allStores {
+		if _, ok := aliveStoreIDsInTiDBZone[id]; ok {
+			storesMatchedPolicy = append(storesMatchedPolicy, id)
+		}
+	}
+	// If no available stores in TiDB zone, we need to access TiFlash stores in other zones.
+	if len(storesMatchedPolicy) == 0 {
+		// needsCrossZoneAccess indicates whether we need to access(directly read or remote read) TiFlash stores in other zones.
+		needsCrossZoneAccess = true
+
+		if policy == tiflash.ClosestAdaptive {
+			// If the policy is `ClosestAdaptive`, we can dispatch tasks to the TiFlash stores in other zones.
+			storesMatchedPolicy = allStores
+		} else if policy == tiflash.ClosestReplicas {
+			// If the policy is `ClosestReplicas`, we dispatch tasks to the TiFlash stores in TiDB zone and remote read from other zones.
+			for id := range aliveStoreIDsInTiDBZone {
+				storesMatchedPolicy = append(storesMatchedPolicy, id)
+			}
+		}
+	}
+	return
+}
+
+// getAliveStoresAndStoreIDs gets all alive TiFlash stores and their IDs.
+// If `careTiFlashReplicaRead` is true, it will also return the IDs of the alive TiFlash stores in TiDB zone.
+func getAliveStoresAndStoreIDs(ctx context.Context, cache *RegionCache, ttl time.Duration, store *kvStore, careTiFlashReplicaRead bool, tidbZone string) (aliveStores []*tikv.Store, aliveStoreIDsInAllZones, aliveStoreIDsInTiDBZone map[uint64]struct{}) {
+	allTiFlashStores := cache.RegionCache.GetTiFlashStores(tikv.LabelFilterNoTiFlashWriteNode)
```

Comment:
This will trigger a pd rpc, is it expected?

Reply:
Yes, it also triggered a pd rpc before.

https://github.com/pingcap/tidb/pull/44107/files#diff-dec6808c7659579faf2c9c6f9d8b7a7e0d7095b61b9a07cfd2cbc9546a3c7c86L332